dotnet/src/Extensions/Planning.StepwisePlanner/StepwisePlanner.cs

```diff
@@ -98,95 +99,286 @@ public Plan CreatePlan(string goal)
+        var textCompletion = aiService as ITextCompletion;
+
+        string userManual = await this.GetUserManual(context).ConfigureAwait(false);
+        string userQuestion = await this.GetUserQuestion(context).ConfigureAwait(false);
 
-            context.Variables.Update($"Result not found, review _stepsTaken to see what happened.\n{JsonSerializer.Serialize(stepsTaken)}");
+#pragma warning disable CA2016 // Forward the 'CancellationToken' parameter to methods that take one -- the method is obsolete and to be removed.
+        var systemContext = this._kernel.CreateNewContext();
+#pragma warning restore CA2016
+
+        systemContext.Variables.Set("suffix", this.Config.Suffix);
+
+        if (chatCompletion is null)
+        {
+            systemContext.Variables.Set("functionDescriptions", userManual);
+        }
+        string systemMessage = await this.GetSystemMessage(systemContext).ConfigureAwait(false);
+
+        chatHistory.AddSystemMessage(systemMessage);
+        if (chatCompletion is not null)
+        {
+            chatHistory.AddUserMessage(userManual);
+        }
+        chatHistory.AddUserMessage(userQuestion);
+
+        return chatHistory;
+    }
+
+    private ChatHistory CreateChatHistory(IKernel kernel, out IAIService aiService)
+    {
+        ChatHistory chatHistory;
+        if (TryGetChatCompletion(this._kernel, out var chatCompletion))
+        {
+            chatHistory = chatCompletion.CreateNewChat();
+            aiService = chatCompletion;
         }
         else
         {
-            context.Variables.Update("Question not found.");
+            var textCompletion = this._kernel.GetService<ITextCompletion>();
+            aiService = textCompletion;
+            chatHistory = new ChatHistory();
         }
 
-        return context;
+        return chatHistory;
     }
 
+    private static bool TryGetChatCompletion(IKernel kernel, [NotNullWhen(true)] out IChatCompletion? chatCompletion)
+    {
+        try
+        {
+            // Client used to request answers to chat completion models
+            chatCompletion = kernel.GetService<IChatCompletion>();
+            return true;
+        }
+        catch (SKException)
+        {
+            chatCompletion = null;
+        }
+
+        return false;
+    }
+
+    private Task<string> GetUserManual(SKContext context)
+    {
+        context.Variables.Set("functionDescriptions", this.GetFunctionDescriptions());
+        return this._promptRenderer.RenderAsync(this._manualTemplate, context);
+    }
+
+    private Task<string> GetUserQuestion(SKContext context)
+    {
+        return this._promptRenderer.RenderAsync(this._questionTemplate, context);
+    }
+
+    private Task<string> GetSystemMessage(SKContext context)
+    {
+        return this._promptRenderer.RenderAsync(this._promptTemplate, context);
+    }
+
+    #endregion setup helpers
+
+    #region execution helpers
     public virtual SystemStep ParseResult(string input)
```

Comment:
Does this really need to be public? We should rename to make it more clear what it's doing and add some xml comments if so.

Reply:
making this protected internal